classes/external/search_packages.php

```diff
@@ -0,0 +1,399 @@
+        $sql = implode(' OR ', $parts);
+        return [$sql, $params];
+    }
+
+    /**
+     * Retrieves the tags and package versions of multiple packages given their ids.
+     *
+     * @param int[] $ids Package ids.
+     * @return array[] A list containing the tags and versions.
+     * @throws moodle_exception
+     */
+    private static function get_tags_and_versions(array $ids): array {
+        global $DB, $USER;
+
+        if (empty($ids)) {
+            return [[], []];
+        }
+
+        // Create sql fragment and parameters.
+        [$insql, $inparams] = $DB->get_in_or_equal($ids);
+
+        // Get tags.
+        $tagsraw = $DB->get_records_sql("
+            SELECT id, packageid, tag
+            FROM {qtype_questionpy_tags}
+            WHERE packageid {$insql}
+        ", $inparams);
+
+        $tags = [];
+        foreach ($tagsraw as $tag) {
+            $tags[$tag->packageid][] = [
+                'id' => $tag->id,
+                'tag' => $tag->tag,
+            ];
+        }
+
+        // Get package versions.
+        $versionsraw = $DB->get_records_sql("
+            SELECT id, packageid, hash, version, userid
+            FROM {qtype_questionpy_pkgversion}
+            WHERE packageid {$insql}
+        ", $inparams);
+
+        $versions = [];
+        foreach ($versionsraw as $version) {
+            $versions[$version->packageid][] = [
+                'id' => $version->id,
+                'hash' => $version->hash,
+                'version' => $version->version,
+                'ismine' => $version->userid === $USER->id,
+            ];
+        }
+
+        return [$tags, $versions];
+    }
+
+    /**
+     * Constructs 'ORDER BY' sql fragment.
+     *
+     * @param string $sort Kind of ordering {@see search_packages::SORT}.
+     * @param string $order Direction of the ordering {@see search_packages::ORDER}.
+     * @return string The sql fragment.
+     */
+    private static function create_order_by_sql(string $sort, string $order): string {
+        $sql = 'ORDER BY';
+        if ($sort === 'alpha') {
+            $sql .= ' name';
+        } else {
+            $sql .= ' timecreated';
+        }
+        return $sql . ' ' . $order;
+    }
+
+    /**
+     * Constructs sql fragment for filtering by tags.
+     *
+     * TODO: change the logic when tags are localized.
+     *
+     * @param array $tags
+     * @return array A list containing the constructed sql fragment, the conditions and an array of parameters.
+     */
+    private static function create_tag_filter_sql(array $tags): array {
+        $joinsql = '';
+        $params = [];
+        $wheretags = [];
+        foreach ($tags as $i => $tag) {
+            $jointagsparam = "tag$i";
+            $params[$jointagsparam] = $tag;
+            $joinsql .= "
+                LEFT JOIN {qtype_questionpy_tags} t$i
+                ON t$i.packageid = p.id AND t$i.id = :$jointagsparam
+            ";
+            $wheretags[] = "NOT t$i IS NULL";
```

Comment:
Nur eine Kleinigkeit, ich würde hier `t$i IS NOT NULL` schreiben. Wenn ich mich nicht irre, sollte diese Where-Bedingung aber überflüssig werden, wenn du statt eines `LEFT JOIN` ein `JOIN` verwendest.